kernel: workq: queue: unify APis

tyrantlizard77 opened the pull request:
We can just do a runtime check for submitting to a workqueue,
it doesn't depend on whether the target queue is running in
user mode or not.

Similar situation with queue objects, an allocation is now implicit when called from user mode. The old alloc APIs are now deprecated.

Commit:
kernel: workq: queue: unify alloc APis

We can just do a runtime check for submitting to a workqueue,
it doesn't depend on whether the target queue is running in
user mode or not.

For queues, we implictly do a resource pool allocation if
called from user mode, skipping it if not. It doesn't matter
from the receiving end.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross reviewed (dismissed):
Love it

andrewboie commented on include/kernel.h:
```diff
@@ -2545,52 +2553,28 @@ static inline void k_work_init(struct k_work *work, k_work_handler_t handler)
+		} else
+#endif /* CONFIG_USERSPACE */
+		{
+			k_queue_append(&work_q->queue, work);
```
Haven't seen anything like that in the MISRA spec I have.

ceolin replied:
actually IS_ENABLED that causes problems. I didn't see anything against extra braces. IS_ENABLE expand for constant expressions that is a problem according with MISRA-C

Commit:
kernel: fix z_thread_malloc() from ISR

ISRs are not in the context of any particular thread and should
not be able to make allocations from the "current" thread's
resource pool.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Hold off on merging this -- I think I can do even better by unifying the k_queue_alloc_append and k_queue_append APIs

dbkinder reviewed (approved):
+1 for doc changes

andrewboie commented:
Adding DNM label -- doing some testing to ensure I haven't missed any cases where this could be a security issue.

andrewboie commented:
OK, I need to close this. I can't make the allocation decision for k_queue based on what mode the caller is running in. This introduces a security problem for the case where we are working with queues with data that user mode threads have access.

The problem is when a supervisor thread enqueues an item. Any user threads consuming that item will have access to the entire data item. With the variants that do an allocation, this is no problem because the queue linked list is allocated in private memory and user threads which might have access to the item won't be able to tamper with the linked list pointers before it gets removed from the queue.

But if you don't allocate, then the private kernel pointers are stored within the item itself. While it sits in the queue, before it is consumed, this information could be tampered with. So we really need to provide supervisor mode with the ability to also enqueue with an allocation.

This transitively affects adding to a workqueue whose work items have user mode access, these also need to always be enqueued with an allocation for the same reasons above.
